service/worker/archiver/history_blob_downloader.go

```diff
@@ -0,0 +1,150 @@
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package archiver
+
+import (
+	"context"
+	"encoding/json"
+	"github.com/uber/cadence/common"
+	"github.com/uber/cadence/common/blobstore"
+	"github.com/uber/cadence/common/blobstore/blob"
+)
+
+type (
+	// DownloadBlobRequest is request to DownloadBlob
+	DownloadBlobRequest struct {
+		NextPageToken        []byte
+		ArchivalBucket       string
+		DomainID             string
+		WorkflowID           string
+		RunID                string
+		CloseFailoverVersion *int64
+	}
+
+	// DownloadBlobResponse is response from DownloadBlob
+	DownloadBlobResponse struct {
+		NextPageToken []byte
+		HistoryBlob   *HistoryBlob
+	}
+
+	// HistoryBlobDownloader is used to download history blobs
+	HistoryBlobDownloader interface {
+		DownloadBlob(context.Context, *DownloadBlobRequest) (*DownloadBlobResponse, error)
+	}
+
+	historyBlobDownloader struct {
+		blobstoreClient blobstore.Client
+	}
+
+	archivalToken struct {
+		BlobstorePageToken   int
+		CloseFailoverVersion int64
+	}
+)
+
+// NewHistoryBlobDownloader returns a new HistoryBlobDownloader
+func NewHistoryBlobDownloader(blobstoreClient blobstore.Client) HistoryBlobDownloader {
+	return &historyBlobDownloader{
+		blobstoreClient: blobstoreClient,
+	}
+}
+
+// DownloadBlob is used to access a history blob from blobstore.
+// CloseFailoverVersion can be optionally provided to get a specific version of a blob, if not provided gets highest available version.
+func (d *historyBlobDownloader) DownloadBlob(ctx context.Context, request *DownloadBlobRequest) (*DownloadBlobResponse, error) {
+	var token *archivalToken
+	var err error
+	if request.NextPageToken != nil {
+		token, err = deserializeArchivalToken(request.NextPageToken)
+		if err != nil {
+			return nil, err
+		}
+	} else if request.CloseFailoverVersion != nil {
+		token = &archivalToken{
+			BlobstorePageToken:   common.FirstBlobPageToken,
+			CloseFailoverVersion: *request.CloseFailoverVersion,
+		}
+	} else {
+		indexKey, err := NewHistoryIndexBlobKey(request.DomainID, request.WorkflowID, request.RunID)
+		if err != nil {
+			return nil, err
+		}
+		indexTags, err := d.blobstoreClient.GetTags(ctx, request.ArchivalBucket, indexKey)
+		if err != nil {
+			return nil, err
+		}
+		highestVersion, err := GetHighestVersion(indexTags)
+		if err != nil {
+			return nil, err
+		}
+		token = &archivalToken{
+			BlobstorePageToken:   common.FirstBlobPageToken,
+			CloseFailoverVersion: *highestVersion,
+		}
+	}
+	key, err := NewHistoryBlobKey(request.DomainID, request.WorkflowID, request.RunID, token.CloseFailoverVersion, token.BlobstorePageToken)
+	if err != nil {
+		return nil, err
+	}
+	b, err := d.blobstoreClient.Download(ctx, request.ArchivalBucket, key)
+	if err != nil {
+		return nil, err
+	}
+	unwrappedBlob, wrappingLayers, err := blob.Unwrap(b)
+	if err != nil {
+		return nil, err
+	}
+	historyBlob := &HistoryBlob{}
+	switch *wrappingLayers.EncodingFormat {
```

Comment:
How about adding a default case and returns an error? I understand that right now there's only one encoding, but if we add a new one and forget to make a change here, an error should be returned.